File: M3U8Joiner/src/dwl/load/pool.go
package load

import (
	"dwl/list"
	"dwl/settings"
	"sync"
	"time"
)

type Pool struct {
	workers []*Worker
	list    *list.List
	sets    *settings.Settings
	mut     sync.Mutex
	working bool
	wait    chan bool
	err     error
}

type Worker struct {
	index  int
	list   *list.List
	sets   *settings.Settings
	iswork bool
	status int
	err    error
	update func()
}

func NewPool(sets *settings.Settings, list *list.List) *Pool {
	p := new(Pool)
	p.sets = sets
	p.list = list
	p.wait = make(chan bool, 1)
	return p
}

func NewWorker(i int, update func(), list *list.List, sets *settings.Settings) *Worker {
	w := new(Worker)
	w.update = update
	w.list = list
	w.sets = sets
	w.index = i
	return w
}

func (p *Pool) Push(itm int, update func()) bool {
	if itm == -1 || p.workers == nil {
		return false
	}
	p.mut.Lock()
	defer p.mut.Unlock()
	if len(p.workers) >= p.sets.Threads {
		return false
	}
	for _, w := range p.workers {
		if w.index == itm {
			return false
		}
	}
	p.workers = append(p.workers, NewWorker(itm, update, p.list, p.sets))
	return true
}

func (p *Pool) Len() int {
	return len(p.workers)
}

func (p *Pool) Start() {
	p.mut.Lock()
	if p.working {
		p.mut.Unlock()
		return
	}
	p.working = true
	p.err = nil
	p.workers = make([]*Worker, 0)
	p.mut.Unlock()

	go func() {
		done := make(chan struct{}, 64)
		for p.working {
			p.mut.Lock()
			isLoad := false
			for i := 0; i < len(p.workers); i++ {
				w := p.workers[i]
				if w.status == STATUS_ERROR {
					p.err = w.err
					p.working = false
					break
				} else if w.status == STATUS_COMPLETE {
					p.workers = append(p.workers[:i], p.workers[i+1:]...)
					i--
					continue
				} else if w.status == STATUS_STOPED && p.working {
					isLoad = true
					go func() {
						w.start()
						done <- struct{}{}
					}()
				}
			}
			p.mut.Unlock()
			if isLoad {
				<-done
			}
		}
		p.wait <- true
	}()
}

func (p *Pool) Stop() {
	p.working = false
	p.mut.Lock()
	defer p.mut.Unlock()
	if len(p.workers) > 0 {
		for _, w := range p.workers {
			w.stop()
		}
		<-p.wait
	}
}

func (p *Pool) Error() error {
	return p.err
}

func (w *Worker) start() {
	w.status = STATUS_LOADING
	w.err = nil
	for i := 0; i < w.sets.ErrorRepeat && w.status == STATUS_LOADING; i++ {
		w.err = w.list.Load(w.index)
		if w.err == nil {
			w.status = STATUS_COMPLETE
			break
		}
		time.Sleep(time.Second)
	}
	if w.err != nil {
		w.status = STATUS_ERROR
	}
	if w.update != nil {
		w.update()
	}
}

func (w *Worker) stop() {
	if w.status != STATUS_COMPLETE {
		w.status = STATUS_STOPED
	}
	if itm := w.list.Get(w.index); itm != nil {
		itm.SetLoading(false)
	}
}

File: m3u8loader/src/main/java/ru/yourok/dwl/downloader/Pool.kt
package ru.yourok.dwl.downloader

import ru.yourok.dwl.settings.Settings
import java.net.SocketException
import java.net.SocketTimeoutException
import kotlin.concurrent.thread


class Pool(private val workers: List<Pair<Worker, DownloadStatus>>) {

    @Volatile
    private var stop = true
    private var currentWorker = 0
    private var thread: Thread? = null
    private var lock: Any = Any()
    private var error: String = ""

    private var onEnd: (() -> Unit)? = null
    private var onFinish: (() -> Unit)? = null
    private var onError: ((err: String) -> Unit)? = null

    fun onEnd(onEnd: () -> Unit) {
        this.onEnd = onEnd
    }

    fun onFinishWorker(onFinish: () -> Unit) {
        this.onFinish = onFinish
    }

    fun onError(onError: (err: String) -> Unit) {
        this.onError = onError
    }

    fun start() {
        stop = false

        this.thread = thread {
            try {
                currentWorker = 0
                var priorityIndex = -1
                workers.forEach { item ->
                    val wrk = item.first
                    val dstat = item.second
                    if (stop || !error.isEmpty())
                        return@forEach
                    synchronized(lock) {
                        currentWorker++
                    }
                    thread {
                        Thread.currentThread().priority = Thread.MIN_PRIORITY
                        for (i in 0..Settings.errorRepeat)
                            try {
                                if (!wrk.item.isComplete && !stop) {
                                    wrk.run()
                                    Runtime.getRuntime().gc()
                                    dstat.isError = false
                                    onFinish?.invoke()
                                }
                                break
                            } catch (e: SocketException) {
                                dstat.isError = true
                                if (i == Settings.errorRepeat) {
                                    error = (e.message
                                            ?: "Error, read or connect") + " " + wrk.item.url + " " + wrk.item.index
                                    onError?.invoke(error)
                                }
                            } catch (e: SocketTimeoutException) {
                                dstat.isError = true
                                if (i == Settings.errorRepeat) {
                                    error = "Error, connection timeout on load item " + wrk.item.index
                                    onError?.invoke(error)
                                }
                            } catch (e: Exception) {
                                e.printStackTrace()
                                dstat.isError = true
                                error = "Error: " + e.message ?: "unknown error"
                                onError?.invoke(error)
                            }
                        synchronized(lock) {
                            currentWorker--
                        }
                    }
                    priorityIndex = setPrior(priorityIndex)
                    while (currentWorker >= Settings.threads)
                        Thread.sleep(100)
                }

                while (currentWorker > 0)
                    Thread.sleep(100)
                onEnd?.invoke()
                stop = true
            } catch (e: Exception) {
                onEnd?.invoke()
                stop = true
                if (!error.isEmpty())
                    onError?.invoke(e.message ?: "Error load items")
            }
        }
    }

    fun stop() {
        stop = true
        for ((wrk, _) in workers)
            wrk.stop()
        thread?.join()
    }

    fun waitEnd() {
        thread?.join()
    }

    fun size(): Int {
        return currentWorker
    }

    fun isWorking(): Boolean {
        return !stop || currentWorker > 0
    }

    private fun setPrior(lastIndex: Int): Int {
        if (lastIndex == -1) {//Find and set
            workers.forEachIndexed { index, pair ->
                if (pair.second.isLoading) {
                    pair.first.setMaxPrior()
                    return index
                }
            }
            return -1
        }
        if (lastIndex in 0..workers.size && workers[lastIndex].second.isCompleteLoad)
            return -1
        return lastIndex
    }
}